[SPARK-14497][ML] Use top instead of sortBy() to get top N frequent words as dict in ConutVectorizer

lionelfeng opened the pull request:
## What changes were proposed in this pull request?

Replace sortBy() with top() to calculate the top N frequent words as dictionary.
## How was this patch tested?

existing unit tests.  The terms with same TF would be sorted in descending order. The test would fail if hardcode the terms with same TF the dictionary like "c", "d"...

lionelfeng committed:
use top() insteand sort() to get dict

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -152,16 +152,10 @@ class CountVectorizer(override val uid: String)
-      tmpSortedWC.map(_._1)
-    }
+
+    val vocab: Array[String] = wordCounts.map(_.swap)
```
Rather than swap them, you can provide the ordering directly to top, right?
Nit: math.min instead of Math.min.

lionelfeng committed:
add custom ordering for top to avoid swap and sorting terms

lionelfeng committed:
return 0 in the comparitor

lionelfeng commented:
cc @lresende

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -152,17 +152,17 @@ class CountVectorizer(override val uid: String)
     }
 
+    val vocab: Array[String] = wordCounts
+      .top(math.min(fullVocabSize, vocSize).toInt)(wordCountOrdering)
```
`top(fullVocabSize, vocSize).toInt)(Ordering.by(_._2))`

mengxr commented:
ok to test

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55437/
Test FAILed.

lresende commented:
Jenkins retest this please

SparkQA commented:
**[Test build #55438 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55438/consoleFull)** for PR 12265 at commit [`af2c3a8`](https://github.com/apache/spark/commit/af2c3a890787ef111da48220d66859637168951f).

SparkQA commented:
**[Test build #55438 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55438/consoleFull)** for PR 12265 at commit [`af2c3a8`](https://github.com/apache/spark/commit/af2c3a890787ef111da48220d66859637168951f).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55438/
Test FAILed.

lresende commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -152,17 +152,17 @@ class CountVectorizer(override val uid: String)
-        wordCounts.sortBy(_._2, ascending = false).take(vocSize)
+
+    val wordCountOrdering = new Ordering[(String, Long)]{
+      override def compare(x: (String, Long), y: (String,Long)): Int = {
```
Please fix scala style issue ==> y: (String, Long))

lionelfeng committed:
replace it with Ordering.By(_._2)

SparkQA commented:
**[Test build #55439 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55439/consoleFull)** for PR 12265 at commit [`38d7c87`](https://github.com/apache/spark/commit/38d7c876d81de87a9c51e1f63dcb4d288ff02955).

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -152,16 +152,10 @@ class CountVectorizer(override val uid: String)
-    }
+
+    val vocab: Array[String] = wordCounts.
+    top(math.min(fullVocabSize, vocSize).toInt)(Ordering.by(_._2))
```
@lionelfeng Please use the following style (moving `.` to next line and fix indentation):

~~~scala
val vocab: Array[String] = wordCounts
  .top(math.min(fullVocabSize, vocSize).toInt)(Ordering.by(_._2))
  .map(_._1)
~~~

SparkQA commented:
**[Test build #55439 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55439/consoleFull)** for PR 12265 at commit [`38d7c87`](https://github.com/apache/spark/commit/38d7c876d81de87a9c51e1f63dcb4d288ff02955).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55439/
Test FAILed.

lionelfeng commented:
Can I change the test a bit since the order of terms with same frequency shouldn't matter in the vocabulary. 

```scala
test("CountVectorizer common cases") {
    val df = sqlContext.createDataFrame(Seq(
      (0, split("a b c d d e"),
        Vectors.sparse(5, Seq((0, 1.0), (1, 1.0), (2, 1.0), (3, 2.0), (4, 1.0)))),
      (1, split("a a a a a a"), Vectors.sparse(5, Seq((0, 6.0)))),
      (2, split("c"), Vectors.sparse(5, Seq((2, 1.0)))),
      (3, split("b b b b b"), Vectors.sparse(5, Seq((1, 5.0)))))
    ).toDF("id", "words", "expected")
    val cv = new CountVectorizer()
      .setInputCol("words")
      .setOutputCol("features")
      .fit(df)
    assert(cv.vocabulary === Array("a", "b", "c", "d", "e"))

    cv.transform(df).select("features", "expected").collect().foreach {
      case Row(features: Vector, expected: Vector) =>
        assert(features ~== expected absTol 1e-14)
    }
  }

```

lionelfeng committed:
avoid validating the order of terms with same frequecy in vocalbulary

SparkQA commented:
**[Test build #55442 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55442/consoleFull)** for PR 12265 at commit [`f6643c2`](https://github.com/apache/spark/commit/f6643c2d7734ecace044b043102328422f46254c).

SparkQA commented:
**[Test build #55442 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55442/consoleFull)** for PR 12265 at commit [`f6643c2`](https://github.com/apache/spark/commit/f6643c2d7734ecace044b043102328422f46254c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55442/
Test PASSed.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -152,16 +152,10 @@ class CountVectorizer(override val uid: String)
-    }
+
+    val vocab: Array[String] = wordCounts
+    .top(math.min(fullVocabSize, vocSize).toInt)(Ordering.by(_._2))
```
These two lines need to be indented. You also don't need a variable type on `vocab`. I think the rest is good.

lionelfeng committed:
indent lines and remove type for vocab.

SparkQA commented:
**[Test build #55462 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55462/consoleFull)** for PR 12265 at commit [`5badc2f`](https://github.com/apache/spark/commit/5badc2f2be915b75853fb40021cd7aa433aa50dc).

SparkQA commented:
**[Test build #55462 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55462/consoleFull)** for PR 12265 at commit [`5badc2f`](https://github.com/apache/spark/commit/5badc2f2be915b75853fb40021cd7aa433aa50dc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55462/
Test FAILed.

lionelfeng committed:
added a row 'd' in test case to make sure no term has the same frequency

SparkQA commented:
**[Test build #55464 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55464/consoleFull)** for PR 12265 at commit [`14d9865`](https://github.com/apache/spark/commit/14d9865e42d01a3886a77d632d9b0ddd13c4e7e1).

SparkQA commented:
**[Test build #55464 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55464/consoleFull)** for PR 12265 at commit [`14d9865`](https://github.com/apache/spark/commit/14d9865e42d01a3886a77d632d9b0ddd13c4e7e1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55464/
Test PASSed.

mengxr commented:
LGTM. Merged into master. Thanks!
